<a href="https://colab.research.google.com/github/jrhulteen/hello-world/blob/master/Notebook2Hulteen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 2

Name: Jacob Hulteen

In [153]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [154]:
#import the data:

#make sure the path on the line below corresponds to the path where you put your dataset.
data = pd.read_csv('/content/drive/MyDrive/CS167/vehicles.csv')
data

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


barrels08  barrelsA08  charge120  ...  phevCity  phevHwy  phevComb
0      15.695714         0.0        0.0  ...         0        0         0
1      29.964545         0.0        0.0  ...         0        0         0
2      12.207778         0.0        0.0  ...         0        0         0
3      29.964545         0.0        0.0  ...         0        0         0
4      17.347895         0.0        0.0  ...         0        0         0
...          ...         ...        ...  ...       ...      ...       ...
42098  14.982273         0.0        0.0  ...         0        0         0
42099  14.330870         0.0        0.0  ...         0        0         0
42100  15.695714         0.0        0.0  ...         0        0         0
42101  15.695714         0.0        0.0  ...         0        0         0
42102  18.311667         0.0        0.0  ...         0        0         0

[42103 rows x 83 columns]

## Exercise #1: 
Work with the right subset: You're not going to work with the whole data set, just the cars with "Regular" listed in the "fuelType" column. So, to start, make a subset of the original data.

In [170]:
regular = data[data['fuelType']=='Regular']
regular1 = regular[['comb08', 'year', 'cylinders','displ']]
regular1

comb08  year  cylinders  displ
0          21  1985        4.0    2.0
1          11  1985       12.0    4.9
2          27  1985        4.0    2.2
3          11  1985        8.0    5.2
5          22  1993        4.0    1.8
...       ...   ...        ...    ...
42097      25  1993        4.0    1.9
42098      22  1993        4.0    2.2
42099      23  1993        4.0    2.2
42100      21  1993        4.0    2.2
42101      21  1993        4.0    2.2

[26863 rows x 4 columns]

## Exercise #2:

Check for null values in the target and predictor columns. If you have a null target value, you will need to throw that example out. If you have a null predictor value, you can either fill them in with something (like the mean/median) or you can drop those rows from the data set. Describe in a markup cell what you decided to do with the null data.

In [181]:
#Replacing the empty values in cylinders and displ with the mean values
avg_cylinders = regular1['cylinders'].mean()
avg_displ = regular1['displ'].mean()
data1 = regular1.copy()
data1['cylinders'] = regular1['cylinders'].fillna(avg_cylinders)
data1['displ'] = regular1['displ'].fillna(avg_displ)
data1.isna().any()

comb08       False
year         False
cylinders    False
displ        False
dtype: bool

*Took the empty cells and replaced them with the average of the rest of the column*

## Exercise #3:

Write up a k-nearest-neighbors function like the one you made for the iris data set in class. It should be able to make mpg ("comb08") predictions for new cars based on the year, cylinders, and displacement. You should also be able to specify what you want to use as k.

In [182]:
import numpy
def knn(predictor, data1, k):
   #calculate the distance
  data1['distance_to_new'] = numpy.sqrt(
    (predictor['year']-data1['year'])**2
    +(predictor['cylinders']-data1['cylinders'])**2
    +(predictor['displ']-data1['displ'])**2)
  #sort by 'distance'
  sorted_data = data1.sort_values(['distance_to_new']) 
  #return the mean() of the k closest neighbors
  return sorted_data.iloc[0:k]['comb08'].mean()

## Exercise #4:

Demonstrate that your function works by making up some new values for hypothetical cars and using your function to display the predicted mpg ("comb08") for that car.

In [183]:
testcar = {}
testcar['year']=1990
testcar['cylinders']=4.0
testcar['displ']=2.0
#Running function
knn(testcar, data1,3)

19.666666666666668

## Exercise #5:

Make a copy of the data and normalize the training data using Z-score. Predict the mpg of a particular car using your k-nearest-neighbors function with both the normalized and non-normalized training data. Compare your results. Use a markup cell to describe and explain the differences in a few sentences.

In [185]:
data2 = data1.copy()
#Calculating mean
year_mean = data2['year'].mean()
cylinders_mean = data2['cylinders'].mean()
displ_mean = data2['displ'].mean()
#Calculating std
year_std = data2['year'].std()
cylinders_std = data2['cylinders'].std()
displ_std = data2['displ'].std()
#Replacing columns with z-score
data2['year'] = (data2['year']-year_mean)/year_std
data2['cylinders'] = (data2['cylinders']-cylinders_mean)/cylinders_std
data2['displ'] = (data2['displ']-displ_mean)/displ_std
#Normalize testcar
testcar['year'] = (1990-year_mean)/year_std
testcar['cylinders'] = (4.0-cylinders_mean)/cylinders_std
testcar['displ'] = (2.0-displ_mean)/displ_std
knn(testcar, data2, 3)

20.666666666666668

For the non-normalized data, using the parameters of year=1990, cylinders=4.0, and displ=2.0, I got 19.67 mpg using 3 nearest neighbors. Using the same parameters but normalizing the data, I got 20.67 mpg using 3 nearest neighbors.  The non-normalized data used the actual data entries of year, cylinders, and displacement which wasn't horribly wrong for this dataset due to the fact they're all numeric to start with, but year was a relatively different numeric value than cylinders and displacement.  As for the normalized data where I calculated the z-score for year, cylinders, and displacement, the values were weighted more similar to one another, causing the mpg to change from 19.67 (non-normalized) to 20.67 (normalized).